Sources/TLS/Socket.swift

```diff
@@ -1,174 +1,176 @@
+    public init(_ context: Context, _ socket: TCPInternetSocket) throws {
+        self.context = context
         self.socket = socket
     }
     
-    public var currSocket: TCPInternetSocket?
-    public var currContext: OpaquePointer?
-    public var lastError:String? {
-        guard let context = currContext, let reason = tls_error(context) else {
-            return nil
-        }
-        return String(validatingUTF8: reason)
-    }
-    
     public convenience init(
         mode: Mode,
         hostname: String,
         port: UInt16 = 443,
         certificates: Certificates = .defaults,
         verifyHost: Bool = true,
         verifyCertificates: Bool = true,
-        cipher: Config.Cipher = .compat,
-        proto: [Config.TLSProtocol] = [.all]
-        ) throws {
-        let context = try Context(mode: mode)
-        let config = try Config(
-            context: context,
-            certificates: certificates,
+        cipherSuite: String? = nil
+    ) throws {
+        let context = try Context(
+            mode,
+            certificates,
             verifyHost: verifyHost,
             verifyCertificates: verifyCertificates,
-            cipher: cipher,
-            proto: proto
+            cipherSuite: cipherSuite
         )
         
         let address = InternetAddress(hostname: hostname, port: port)
         let socket = try TCPInternetSocket(address: address)
         
-        try self.init(config: config, socket: socket)
+        try self.init(context, socket)
     }
     
     
-    /**
-     Connects to an SSL server from this client.
-     
-     This should only be called if the Context's mode is `.client`
-     */
+    /// Connects to an SSL server from this client.
+    ///
+    /// This should only be called if the Context's mode is `.client`
     public func connect(servername: String) throws {
         try socket.connect()
-        let connectResult = tls_connect_socket(
-            config.context.cContext,
-            socket.descriptor,
-            servername
-        )
-        currSocket = socket
-        currContext = config.context.cContext
-        
-        guard connectResult == Result.OK else {
-            throw TLSError.connect(lastError ?? "Unknown")
+
+        guard let ssl = SSL_new(context.cContext) else {
+            throw makeError(functionName: "SSL_new")
         }
-        
-        // handshake is performed automatically when using tls_read or tls_write, but by doing it here, handshake errors can be properly reported
-        guard tls_handshake(config.context.cContext) == Result.OK else {
-            throw TLSError.handshake(lastError ?? "Unknown")
+        cSSL = ssl
+
+        try assert(
+            SSL_set_fd(ssl, socket.descriptor),
+            functionName: "SSL_set_fd"
+        )
+
+        if context.verifyHost {
+            print("Warning: Host verification not implemented.")
+//            let param = SSL_get0_param(ssl)
```

Comment:
It's in progress. I need to figure out a better fallback for when the openssl / libressl version isn't new enough. It's on my todo.